Sidecar migration

nsklikas opened the pull request:
This PR migrates the gatekeeper manager operator charm to using the latest version of the `ops` library. The old charm was outdated and using the podspec operator pattern. 

We had to patch the StatefulSet to mount a volume to the gatekeeper container in order to make the charm work. This causes the charm operator to restart when first installing the charm.

Some notes/TODOs:
- Add more descriptive README?
- The audit operator will be migrated in a separate PR.
- Fix github workflow
- Do we need more tests?
- This charm uses the built in [opa cert-controller](https://github.com/open-policy-agent/cert-controller), should we allow for the use of cert-manager instead? Perhaps this way we can avoid mounting to the container.

nsklikas committed:
WIP: Migrate to sidecar pattern

This version has several issues: the deployment throws errors
and even after it is deployed and everything seems fine the
policies are not enforced

nsklikas committed:
Remove selectors

Juju is adding its own labels to the pod

nsklikas committed:
Lint code

nsklikas committed:
Fix manager pod deployment

nsklikas committed:
Add config_changed hook

Commit:
Update charm to use manifest lib, get to state where it can deploy

nsklikas committed:
Push uncommitted files

Commit:
Add required patches

Commit:
mock lightkube

Commit:
Patch statefulset

nsklikas committed:
Merge pull request #2 from charmed-kubernetes/scp/migration-updates

Update charm to use manifest lib, get to state where it can deploy

Commit:
Update to latest version of manifest lib

nsklikas committed:
Merge pull request #3 from charmed-kubernetes/scp/update-to-latest-manifest-lib

Update to latest version of manifest lib

nsklikas committed:
Patch Service selector

nsklikas committed:
Working charm version

nsklikas committed:
Add unit tests

Commit:
Remove manifest resources on remove instead of on stop

nsklikas committed:
Merge pull request #4 from charmed-kubernetes/scp/resource-removal-updates

Remove manifest resources on remove instead of on stop

nsklikas committed:
Add list-constraints action

nsklikas committed:
Update integration tests

nsklikas committed:
Format code and add pyproject.toml

nsklikas committed:
Fix PodDisruptionBudget selector

nsklikas committed:
Update README

stonepreston commented on opa-manager-operator/tests/unit/test_charm.py:
```diff
@@ -1,61 +1,67 @@
+    assert service.is_running()
+    assert harness.model.unit.status.name == "active"
+
+    # testing that the dnsPolicy is patched via lightkube
```
```suggestion
    # testing that the statefulset is patched via lightkube
```

stonepreston commented on opa-manager-operator/tests/unit/test_charm.py:
```diff
@@ -1,61 +1,67 @@
+        "services": {
+            "gatekeeper": {
+                "command": "/manager --port=8443 --logtostderr "
+                "--exempt-namespace=gatekeeper-system "
```
this should be the model name instead of gatekeeper-system correct?

nsklikas replied:
you are right

stonepreston commented on opa-manager-operator/tests/unit/test_charm.py:
```diff
@@ -1,61 +1,67 @@
+    assert harness.charm._on_config_changed({}) is None
+
+
+def test_on_stop(harness):
```
I would switch the test to testing the remove event instead of stop. The test should ensure the delete call is made to remove the resources. Mocking out the manifest lib call should be sufficient for this

stonepreston commented on opa-manager-operator/tests/unit/test_charm.py:
```diff
@@ -1,61 +1,67 @@
+
+
+def test_on_config_changed(harness):
+    assert harness.charm._on_config_changed({}) is None
```
this test should ensure the container is restarted

stonepreston commented on opa-manager-operator/tests/unit/test_charm.py:
```diff
@@ -1,61 +1,67 @@
+    patch.assert_called_once()
+
+
+def test_coredns_pebble_ready_already_started(harness, active_container, caplog):
```
```suggestion
def test_gatekeeper_pebble_ready_already_started(harness, active_container, caplog):
```

stonepreston commented on opa-manager-operator/tests/unit/test_charm.py:
```diff
@@ -1,61 +1,67 @@
+import ops.testing
+
+ops.testing.SIMULATE_CAN_CONNECT = True
+
```
An install event test should also be added to ensure the manifests are applied, and that the manipulations worked as intended (ie the deployments from the yaml were skipped, etc)

stonepreston commented on opa-manager-operator/tests/unit/conftest.py:
```diff
@@ -0,0 +1,47 @@
+
+
+@pytest.fixture
+def harness(mocker):
```
I think you will need to include [Pytest-mock](https://pypi.org/project/pytest-mock/) in the test dependencies to use the mocker fixture.

stonepreston commented on opa-manager-operator/tests/integration/test_charm.py:
```diff
@@ -1,115 +1,144 @@
+                '"check-ignore-label.gatekeeper.sh"'
             )
+            # We sleep in order to wait for opa to register the constraint
+            # TODO: Is there a better way to do this?
```
you might be able to use [the lightkube client.wait() method](https://lightkube.readthedocs.io/en/stable/client/) here instead of this loop. If the resource you are wanting to wait on has a ready condition that should work.

nsklikas replied:
The comment was misleading, the problem is we need to wait for the `webhook` to become active. See a related gatekeeper [issue](https://github.com/open-policy-agent/gatekeeper/issues/1156). The gatekeeper helm [hook](https://github.com/open-policy-agent/gatekeeper/blob/master/cmd/build/helmify/static/templates/_helpers.tpl#L51) curls the service until it is up.

The weird thing is that the pebble health check should be enough to ensure when the webhook will be ready and by running `juju status` I can see that the `gatekeeper-controller-manager` app changes from `waiting` to `active` when the webhook is ready and the sleeping stops, but `model.wait_for_idle(status="active")` returns before the application goes `active`. It returns when the units go active instead.

After looking into the `wait_for_idle` code I think that that is the intended behavior, i.e. it checks for the status of the units to determine whether it should return, I guess i will have to find some other way to ensure that the application is `active`.

stonepreston commented on opa-manager-operator/tests/integration/test_charm.py:
```diff
@@ -1,115 +1,144 @@
+    client.delete(Namespace, ns_name)
+
+
+async def test_list_constraints(ops_test):
```
I would also add a corresponding unit test for this action as well.

nsklikas committed:
Clean up

nsklikas committed:
Fix typo

stonepreston commented:
Add more descriptive README?: Good idea
Fix github workflow: Yes definitely move the .github folder up to the project root. We need to ensure the unit and integration tests run for the PR
Do we need more tests?: Yes, I added some review comments with some suggested unit tests. Integration tests look great, but the unit tests are a bit light. 
Re: cert-manager, great idea, although possibly outside the scope of the migration task, however depending on complexity it could be added in in this PR, similar to how prometheus integration is going.

nsklikas committed:
Improve unit test

nsklikas committed:
Fixes

nsklikas committed:
Remove wait_for_idle

Implement and use wait_for_application instead of `wait_for_idle`.
`wait_for_idle` waits for the units of an application to go active,
instead of the actual application. The units may be running, but
until the application is active the webhook is not ready.

nsklikas committed:
Move github workflow to top dir

stonepreston committed:
Add prometheus integration (#6)

* Add prometheus integration

* fix linting

* Fix failing unit tests

* Use reusable workflows

* Add wokeignore

* Remove v0 service patch lib

nsklikas committed:
Update README.md

nsklikas committed:
Remove obsolete file

nsklikas committed:
Add log-level config

nsklikas committed:
Add block state and tests

stonepreston commented on opa-manager-operator/README.md:
```diff
@@ -1,76 +1,58 @@
-
-### Post deployment steps
+After applying this policy all namespaces will need to have the label `gatekeeper`.
+Incomming namespaces are not affected by this.
```
```suggestion
Incoming namespaces are not affected by this.
```

stonepreston commented on opa-manager-operator/README.md:
```diff
@@ -1,76 +1,58 @@
+Incomming namespaces are not affected by this.
 
+### Getting policies
+To list all the policies that are currently being applied run:
```
```suggestion
To list all the policies that are currently applied run:
```

stonepreston commented on opa-manager-operator/actions.yaml:
```diff
@@ -0,0 +1,8 @@
+list-versions:
+  description: List Cloud Provider and Storage Versions supported by this charm
```
```suggestion
  description: List OPA-Gatekeeper versions supported by this charm
```

stonepreston commented on opa-manager-operator/charmcraft.yaml:
```diff
@@ -1,6 +1,28 @@
+        - arm
+        - arm64
+        - ppc64le
+        - s390x
```
```suggestion
```

stonepreston replied:
The [OPA images](https://hub.docker.com/r/openpolicyagent/gatekeeper/tags) only support amd64, arm, and arm64, so thats what we should list here.

stonepreston commented on opa-manager-operator/charmcraft.yaml:
```diff
@@ -1,6 +1,28 @@
+        - arm
+        - arm64
+        - ppc64le
+        - s390x
```
```suggestion
```

stonepreston commented on opa-manager-operator/config.yaml:
```diff
@@ -5,4 +5,12 @@ options:
+    default: v3.9.0
+    description: Version of OPA controller manager to deploy
+    type: string
+  log-level:
```
Nice!

stonepreston commented on opa-manager-operator/src/charm.py:
```diff
@@ -1,172 +1,262 @@
+
+    @property
+    def pod_name(self):
+        # XXX: Temporary hack, we need to somehow get the pod's name
```
```suggestion
```

You can remove the temporary comment, I think this is is a perfectly valid way of getting the pod name

stonepreston commented on opa-manager-operator/src/manifests.py:
```diff
@@ -0,0 +1,141 @@
+    def __call__(self, obj):
+        if obj.metadata.name == "gatekeeper-webhook-service":
+            obj.spec.selector = {
+                "app.kubernetes.io/name": "gatekeeper-controller-manager"
```
It would be nice if we could get the app name dynamically here:
```suggestion
                "app.kubernetes.io/name": self.manifests.model.app.name
```

stonepreston commented on opa-manager-operator/src/manifests.py:
```diff
@@ -0,0 +1,141 @@
+    def __call__(self, obj):
+        if obj.kind == "PodDisruptionBudget":
+            obj.spec.selector.matchLabels = {
+                "app.kubernetes.io/name": "gatekeeper-controller-manager"
```
```suggestion
                "app.kubernetes.io/name": self.manifests.model.app.name
```

stonepreston commented on opa-manager-operator/tests/integration/test_charm.py:
```diff
@@ -1,115 +1,212 @@
+import asyncio
```
Very nice job on the integration tests. Good coverage

stonepreston reviewed (changes requested):
Just a few small things here and there that could use changing,  great job!

nsklikas committed:
Requested fixes

nsklikas committed:
Fix README, remove unused config
